spark/src/main/java/org/apache/iceberg/spark/source/Reader.java

```diff
@@ -308,6 +332,29 @@ private ReadTask(CombinedScanTask task, String tableSchemaString, String expecte
         encryptionManager.value(), caseSensitive);
     }
 
+    @Override
+    public String[] preferredLocations() {
+      if (!enableLocality) {
+        return new String[0];
+      }
+
+      Configuration conf = SparkSession.active().sparkContext().hadoopConfiguration();
+      Set<String> locations = Sets.newHashSet();
+      for (FileScanTask f : task.files()) {
+        Path path = new Path(f.file().path().toString());
+        try {
+          FileSystem fs = path.getFileSystem(conf);
+          for (BlockLocation b : fs.getFileBlockLocations(path, f.start(), f.length())) {
+            locations.addAll(Arrays.asList(b.getHosts()));
```

Comment:
Any chance `getFileBlockLocations ` returns `null` so we'd end up with a NPE for `b.getHosts()`?

Reply:
I checked the Hadoop code, I don't think this will return a null value here, on the contrary, it will return an empty array if there's no hosts.